docs: Update upgrading docs (#14635)
sourcegraph/sourcegraph: doc/admin/updates/docker_compose.md

# Updating a Docker Compose Sourcegraph instance

> This document describes the exact changes needed to update a [Docker Compose Sourcegraph instance](https://github.com/sourcegraph/deploy-sourcegraph-docker/tree/master/docker-compose).

Reviewer: ```suggestion
This document describes the exact changes needed to update a [Docker Compose Sourcegraph deployment](../install/docker-compose.md).
```

I want to move docker compose out of that repo eventually, and the docs page does a better job explaining who should/shouldn't use docker-compose so lets link there instead

Each section comprehensively describes the steps needed to upgrade, and any manual migration steps you must perform.

A new version of Sourcegraph is released every month (with patch releases in between, released as needed). Check the [Sourcegraph blog](https://about.sourcegraph.com/blog) or the site admin updates page to learn about updates. We actively maintain the two most recent monthly releases of Sourcegraph.

Upgrades should happen across consecutive minor versions of Sourcegraph. For example, if you are running Sourcegraph 3.1 and want to upgrade to 3.3, you should upgrade to 3.2 and then 3.3.

**Always refer to this page before upgrading Sourcegraph,** as it comprehensively describes the steps needed to upgrade, and any manual migration steps you must perform.

> ## 3.20.1 -> 3.21.0

Reviewer: Can you add the LSIF migration info under a sub-header like `### If you wish to keep LSIF data` and clarify at the very start what LSIF data is and why you might not have it?

This could allow site admins with no LSIF data to quickly say "Ah, I don't need that" and disregard reading that section while keeping their focus on the part that may be important - like setting the `CODEINTEL_PG` env vars.

Please upgrade to the [`v3.21.0` tag of deploy-sourcegraph-docker](https://github.com/sourcegraph/deploy-sourcegraph-docker/tree/v3.21.0/docker-compose) by following the [standard upgrade procedure](#standard-upgrade-procedure).

This release introduces a second database instance, `codeintel-db`. If you have configured Sourcegraph with an external database, then update the `CODEINTEL_PG*` environment variables to point to a new external database  as described in the [external database documentation](../external_database.md). Again, these must not point to the same database or the Sourcegraph instance will refuse to start.

> If you had LSIF data uploaded prior to upgrading to 3.21.0, there is a background migration that moves all existing LSIF data into the `codeintel-db`. Once this process completes, the `/lsif-storage/dbs` directory on the precise-code-intel-bundle-manager volume should be empty, and the bundle manager should print the following log message:

Reviewer: ```suggestion
If you had LSIF data uploaded prior to upgrading to 3.21.0, there is a background migration that moves all existing LSIF data into the `codeintel-db` upon upgrade. Once this process completes, the `/lsif-storage/dbs` directory on the precise-code-intel-bundle-manager volume should be empty, and the bundle manager should print the following log message:
```

> Migration to Postgres has completed. All existing LSIF bundles have moved to the path /lsif-storage/db-backups and can be removed from the filesystem to reclaim space.

> Once this message has been printed, you are free to delete the bundle files moved into the `/lsif-storage/db-backups` directory on the bundle-manager volume.

Reviewer: ```suggestion
**Wait for the above message to be printed in `docker logs precise-code-intel-bundle-manager` before upgrading to the next Sourcegraph version**, then if everything is working you can free up disk space by deleting the backup bundle files using this command:

```sh
docker exec -it precise-code-intel-bundle-manager sh -c 'rm -rf /lsif-storage/db-backups'
```

> > Warning: In order to ensure there is no data loss, **do not upgrade out of the 3.21.x release branch** until you have seen this log message, or verified that the `/lsif-storage/dbs` directory on the precise-code-intel-bundle-manager volume is empty.

Reviewer: ```suggestion
```

Better to mention this earlier than at the very end. Assume people won't read the whole thing.
